rfcs/0192-min-capacity-ensures-workers-do-not-get-unnecessarily-killed.md

```diff
@@ -0,0 +1,456 @@
+   - VCS repositories and history
+   - Package manager caches (npm, pip, cargo, etc.)
+   - Container images and layers
+
+2. **Provisioning Delays**: New worker provisioning takes ~73-74 seconds average for decision pools, during which tasks must wait
+
+3. **Resource Waste**: The current cycle of shutdown → detection → spawn → provision → register wastes compute resources and increases task latency
+
+4. **Violation of MinCapacity Intent**: `minCapacity >= 1` suggests these pools should always have capacity available, but the current implementation allows temporary capacity gaps
+
+# Details
+
+## Current Behavior Analysis
+
+**Affected Worker Pools:**
+- Direct `minCapacity: 1`: [`infra/build-decision`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L220), [`code-review/bot-gcp`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L3320)
+- Keyed `minCapacity: 1`: [`gecko-1/decision-gcp`, `gecko-3/decision-gcp`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L976), and [pools matching `(app-services|glean|mozillavpn|mobile|mozilla|translations)-1`](https://github.com/mozilla-releng/fxci-config/blob/43c18aab0826244e369b16a964637b6c411c7760/worker-pools.yml#L1088)
+
+**Current Implementation Issues:**
+- Worker-manager enforces minCapacity by spawning new workers when capacity drops below threshold
+- Generic-worker shuts down after `afterIdleSeconds` regardless of minCapacity requirements
+- Gap exists between worker shutdown and replacement detection/provisioning
+
+## Proposed Solution
+
+### Core Concept: MinCapacity Workers
+
+Workers fulfilling `minCapacity >= 1` requirements should:
+1. **Extended Idle Timeout**: Receive significantly longer idle timeouts to preserve caches
+2. **Priority Replacement**: Be immediately replaced when they fail or become unhealthy
+3. **Per-Variant MinCapacity**: Each pool variant (gecko-1 vs gecko-3) maintains its own minCapacity workers
+
+### Implementation Approach
+
+#### 1. Worker Identification and Tagging
+
+**Database Schema Addition:**
+```sql
+ALTER TABLE workers ADD COLUMN intended_role text; -- 'min-capacity' or 'overflow'
+```
+
+**Current Workers Table Schema:** [workers table definition](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/db/tables.yml#L154-L168)
+
+**Worker Provisioning Logic:**
+- Worker-manager determines worker role at spawn time based on current pool needs
+- Workers are launched with predetermined `intended_role`
+- Uses existing [`launch_config_id` system](https://github.com/taskcluster/taskcluster/blob/d7fbf0ee9e0d93e079cc7ff069eaceecfd7d29ec/services/worker-manager/src/data.js#L313-L362) to ensure per-variant capacity tracking
+- **MinCapacity Worker Selection**: When capacity planning, prioritize spawning minCapacity workers first
```

Comment:
i think this is redundant. workers would be created as is to run tasks that are pending, it doesn't matter which order they would be launched in

Reply:
Resolving thread because we agreed on only provisioning permanent workers when we're the number of workers falls below `minCapacity` https://github.com/taskcluster/taskcluster-rfcs/pull/192#discussion_r2382446393